# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/other/RemoteAccessible.cpp
# Commit: 86da8f2a28ea
# Full Hash: 86da8f2a28ea1992671d08dfb56c881a1a82daa5
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-08-04 04:34:13
# Regressor Bug: 1781536
# File Overlap Count: 1
# Description:
#   Bug 1781536 part 3: Move RelationByType to base Accessible. r=morgan
#   
#   This required changing the RemoteAccessible implementations to return Relation instead of an array of RemoteAccessible.
#   Platform implementations have been updated to use the unified method where appropriate.
#   
# ==============================================================================

diff -r 95ae96802dbe -r 86da8f2a28ea accessible/ipc/other/RemoteAccessible.cpp
--- a/accessible/ipc/other/RemoteAccessible.cpp	Wed Aug 03 23:58:51 2022 +0000
+++ b/accessible/ipc/other/RemoteAccessible.cpp	Wed Aug 03 23:58:51 2022 +0000
@@ -13,6 +13,7 @@
 #include "mozilla/dom/BrowserParent.h"
 #include "mozilla/Unused.h"
 #include "mozilla/a11y/Platform.h"
+#include "Relation.h"
 #include "RelationType.h"
 #include "mozilla/a11y/Role.h"
 #include "mozilla/StaticPrefs_accessibility.h"
@@ -76,8 +77,7 @@
   return attrs.forget();
 }
 
-nsTArray<RemoteAccessible*> RemoteAccessible::RelationByType(
-    RelationType aType) const {
+Relation RemoteAccessible::RelationByType(RelationType aType) const {
   if (StaticPrefs::accessibility_cache_enabled_AtStartup()) {
     return RemoteAccessibleBase<RemoteAccessible>::RelationByType(aType);
   }
@@ -85,16 +85,7 @@
   nsTArray<uint64_t> targetIDs;
   Unused << mDoc->SendRelationByType(mID, static_cast<uint32_t>(aType),
                                      &targetIDs);
-
-  size_t targetCount = targetIDs.Length();
-  nsTArray<RemoteAccessible*> targets(targetCount);
-  for (size_t i = 0; i < targetCount; i++) {
-    if (RemoteAccessible* proxy = mDoc->GetAccessible(targetIDs[i])) {
-      targets.AppendElement(proxy);
-    }
-  }
-
-  return targets;
+  return Relation(new RemoteAccIterator(std::move(targetIDs), Document()));
 }
 
 void RemoteAccessible::Relations(